### DISCUSSION TITLE
Is it possible to implement a hybrid search (semantic + full-text) using N8n and Supabase? :(

### DISCUSSION THREAD

**Discussion by Felipe92299229** [supabase/supabase#29712] [Questions]
Hey everyone, I'm trying to build a RAG (Retrieval-Augmented Generation) agent using N8n + Vector Store in Supabase, but I've hit a problem I can't solve. I want to know if it's possible to implement hybrid search (semantic + full-text) using only N8n and Supabase.

I followed the instructions from this link: [Supabase with Langchain](https://supabase.com/docs/guides/ai/langchain?database-method=sql). The example there works well for semantic search, but I also need to combine it with full-text search to achieve more precise results.

What I've tried:

I made some modifications to the original code using GPT to add an fts (full-text search) column and create a hybrid function that combines similarity search with textual relevance search. Unfortunately, I haven't been successful.
I searched for solutions on the web, but everything I've found so far hasn't solved the problem.
My current scenario:

I'm using a table in Supabase to store my documents, with a column for embeddings and metadata.
The search accuracy, even with only 20 chunks of 1000 words, is not satisfactory. For example, when searching for terms like "Chinese pastries" or "chocolate chips," the results are not precise enough.
So, my question is: is it possible to create a hybrid search agent (combining semantic similarity and full-text search) using only N8n and Supabase? If so, does anyone have any example or tips on how to implement this more effectively?

Any help would be greatly appreciated, especially on how to improve the accuracy of the vector store using a hybrid approach. Thanks!

**Comment by jzapataca** [comment]
You have to ensure the hybrid function call three parameters: text, match_count and query embedding, I make some modifications in the function and this was the result (change the size vector based on your embedding model):
```sql
create or replace function public.hybrid_search(
  filter text,
  match_count int,
  query_embedding vector(768)
)
returns setof documents
language sql
as $$
with full_text as (
  select
    id,
    row_number() over(order by ts_rank_cd(fts, websearch_to_tsquery(filter)) desc) as rank_ix
  from
    documents
  where
    fts @@ websearch_to_tsquery(filter)
  order by rank_ix
  limit least(match_count, 30) * 2
),
semantic as (
  select
    id,
    row_number() over (order by embedding <#> query_embedding) as rank_ix
  from
    documents
  order by rank_ix
  limit least(match_count, 30) * 2
)
select
  documents.*
from
  full_text
  full outer join semantic
    on full_text.id = semantic.id
  join documents
    on coalesce(full_text.id, semantic.id) = documents.id
order by
  coalesce(1.0 / (50 + full_text.rank_ix), 0.0) * 1 +
  coalesce(1.0 / (50 + semantic.rank_ix), 0.0) * 1
desc
limit
  least(match_count, 30)
$$;

```
